### PR TITLE
[Defacepocalypse] De-deface variants

### LINKED ISSUE #3615: De-deface admin variants edit page
### What we should change and why (this is tech debt)
We should delete app/overrides/spree/admin/variants and move everything to app/views/spree/admin/variants/_form.
This will require a new angular controller to replace on_demand_script.
There's a bug related to this page that could also be tackled with this: #3249

### Context
Found as we discussed using the new spree v2 stock management page or keep using variants edit page for stock mangement here #3573

### Impact and timeline
I think this is the most complex override we have got right now. Specs should be added to cover this page behaviour.

### PR DESCRIPTION
**Pull request by HugsDaniel**
#### What? Why?

Closes #3615

I imported and de-defaced views from spree/backend/app/views/spree/admin/variants and imported views from Spree. So now we have more views :

`app/views/spree/admin/variants/index.html.haml`
`app/views/spree/admin/variants/_form.html.haml`
`app/views/spree/admin/variants/new.html.haml`
`app/views/spree/admin/variants/edit.html.haml`
`app/views/spree/admin/variants/new.js.erb`

No functional changes done, just bringing code from spree, converting to haml and applying the defaces.

#### What should we test?

Make sure these pages work as before :

 - Product -> Edit -> Variants
 - Product -> Edit -> Variants -> New Variant
 - Product -> Edit -> Variants -> Edit

#### Release notes

Changelog Category: Changed
Removed deface from a few parts of the backoffice thus making OFN less entangled with Spree code.

### CODE REVIEW

**Comment by HugsDaniel** [comment]
Hey @luisramos0 I only imported the form since new and edit were not defaced, do you think I should import them as well ?

**Comment by luisramos0** [comment]
yes, please bring everything that is used in OFN, which is the case for these, we should import these.
Because now it's not just de-defacing, it's removing dependency to spree_backend #4050

**Comment by luisramos0** [comment]
I think there is an issue for this one: https://github.com/openfoodfoundation/openfoodnetwork/issues/3615

**Comment by luisramos0** [comment]
@hugs for your awareness:
https://github.com/openfoodfoundation/openfoodnetwork/issues/1805#issuecomment-515018774

Maybe we improve a few things in this page as part of this PR?

**Comment by daniellemoorhead** [comment]
Note to self: compare this on one staging server with one that doesn't have it.

**Comment by HugsDaniel** [comment]
I moved this back to In Dev because I'm working on adding a ng controller for the on_demand script

**Comment by HugsDaniel** [comment]
@luisramos0 I added my first ng directive o/ It works just fine on my part, can you check it please ?

File: app/assets/javascripts/admin/products/directives/set_variant_on_demand.js.coffee [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    if onDemand.checked
+      onHand.disabled = 'disabled'
+      onHand.dataStock = onHand.value
+      onHand.value = t('admin.products.variants.infinity')
+
+    onDemand.addEventListener 'change', (event) ->
+      if onDemand.checked
+        onHand.disabled = 'disabled'
+        onHand.dataStock = onHand.value
+        onHand.value = t('admin.products.variants.infinity')
```
Comment (HugsDaniel) [inline app/assets/javascripts/admin/products/directives/set_variant_on_demand.js.coffee]: How can I refactor these 3 lines to be on a function ? They are called twice, it's not ideal. I didn't manage to declare a function.

File: app/assets/javascripts/admin/products/directives/set_variant_on_demand.js.coffee [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      else
+        onHand.removeAttribute('disabled')
+        onHand.value = onHand.dataStock
+
+
```
Comment (luisramos0) [inline app/assets/javascripts/admin/products/directives/set_variant_on_demand.js.coffee]: extra empty lines.

File: app/assets/javascripts/admin/products/directives/set_variant_on_demand.js.coffee (hunk 1 above)
Comment (luisramos0) [reply app/assets/javascripts/admin/products/directives/set_variant_on_demand.js.coffee]: I just looked up some other directives in our codebase, I think you can use a named function:


```ruby
angular.module("admin.products").directive "setOnDemand", ->
  link: (scope, element, attr) ->
    onHand = element.context.querySelector("#variant_on_hand")
    onDemand = element.context.querySelector("#variant_on_demand")

     disableOnHandIfOnDemand()

     onDemand.addEventListener 'change', (event) ->
         disableOnHandIfOnDemand()

     disableOnHandIfOnDemand = ->
         if onDemand.checked        
            onHand.disabled = 'disabled'
            onHand.dataStock = onHand.value
            onHand.value = t('admin.products.variants.infinity')
```

File: app/views/spree/admin/variants/_form.html.haml [hunk 3]
```diff
@@ -64,41 +65,3 @@
...
-
-  $(document).ready(function() {
-
-      var on_demand = $('input#variant_on_demand');
-      var on_hand = $('input#variant_on_hand');
-
-      disableOnHandIfOnDemand = function() {
-          on_demand_checked = on_demand.attr('checked')
-          if ( on_demand_checked == undefined )
-              on_demand_checked = false;
```
Comment (luisramos0) [inline app/views/spree/admin/variants/_form.html.haml]: have you verified this is not required? does the new code work when on_demand is null?

File: app/views/spree/admin/variants/_form.html.haml [hunk 4]
```diff
@@ -64,41 +65,3 @@
...
-          }
-      }
-
-      disableOnHandIfOnDemand();
-
-      on_demand.change(function(){
-          disableOnHandIfOnDemand();
-          if(!this.checked) {
-              if(on_hand.attr('data-stock') !== undefined) {
-                  on_hand.val(on_hand.attr('data-stock'));
```
Comment (luisramos0) [inline app/views/spree/admin/variants/_form.html.haml]: are you sure this condition is not necessary?

File: app/assets/javascripts/admin/products/directives/set_variant_on_demand.js.coffee (hunk 1 above)
Comment (HugsDaniel) [reply app/assets/javascripts/admin/products/directives/set_variant_on_demand.js.coffee]: Nice thanks!

**Comment by HugsDaniel** [comment]
@luisramos0 it should be good now !

**Comment by luisramos0** [comment]
significant changes since Maikel's review, so I am requesting re-review ;-)

**Comment by daniellemoorhead** [comment]
Going to test this with Jen on Friday.

**Comment by daniellemoorhead** [comment]
Testing notes: https://docs.google.com/document/d/1nTZtJHF4xoMtXquso-cXGB_UAQxCzLbymf5OgNFgGTk/edit?usp=sharing

All good, ready to go.

**Pull request merged**